src/ml/neural_net/mps_layer_helper.mm

```diff
@@ -9,8 +9,18 @@
 #import <ml/neural_net/mps_layer_instance_norm_data_loader.h>
 #import <ml/neural_net/mps_weight.h>
 
+#import <objc/runtime.h>
+
 @implementation MPSCNNFullyConnectedNode (TCMPSLayerHelper)
-@dynamic weights;
+
+- (void)setWeights:(TCMPSConvolutionWeights *)weights {
+    objc_setAssociatedObject(self, @selector(tag), weights, OBJC_ASSOCIATION_RETAIN);
```

Comment:
The key used for associated objects must be unique. What is `tag` in this case? How can you be sure no one else will use it?

I would recommend the officially endorsed pattern:
```
static char kAssociatedObjectKey;
```
And use `&kAssociatedObjectKey` as the key for these objc functions

Reply:
Ah yes, that `tag` was supposed to be `weights`! I did look into the `kAssociatedObjectKey` pattern according to https://nshipster.com/associated-objects/. It was suggested that the selector pattern was better.

Reply:
AFAICT, `@selector` only gives a unique value if no one else uses the same selector name. So if you use `tag` or `weights`, anyone else who uses the same name could in theory collide. It's safer to use a name that at least includes a TC prefix, but at that point I figure why not go the guaranteed route and use a static variable.

Although maybe it's reasonable to apply the same standard to the category property itself. Maybe it should be `tc_weights` to avoid conflicts with any other category property that somebody else might add. Then I think it would be reasonable to use `@selector(tc_weights)`

Reply:
I'm not sure I understand. Is this a per @instance thing or a per @implementation thing. Would the static variable collide in the multiple implementations?

Reply:
The main thing to understand is that `objc_setAssociatedObject` is just a global function allowing you to specify values for any pair of object instance and "key", where the "key" is just a `void*` value. It's global, so any other code in the same process can call the same function to access the values associated with a given object instance. The safest way to guarantee that some other code doesn't use the same key in a different way is to use a key that no one else knows about. Using `@selector` only satisfies this requirement if the name of the selector is something no one else knows about. A static variable by definition is unique, even if another .m or .cpp file uses the exact same name.